pymc_experimental/model_builder.py

```diff
@@ -0,0 +1,129 @@
+import pymc as pm
+import arviz
+import pandas as pd
+from pathlib import Path
+import numpy as np
+from typing import Dict
+import cloudpickle
+import arviz as az
+import hashlib
+
+
+class ModelBuilder(pm.Model):
+
+    _model_type = "BaseClass"
+    version = "None"
+
+    def __init__(self, data: pd.DataFrame, model_config: Dict, sampler_config: Dict):
+        super().__init__()
+        self.data = data  # input and output data
+        self.model_config = model_config  # parameters for priors etc.
+        self.sample_config = sampler_config  # parameters for sampling
+
+        self.idata = None
+
+    def _build(self):
+        raise NotImplementedError
+
+    def _data_setter(self, data: pd.DataFrame, x_only: bool = True):
+
+        raise NotImplementedError
+
+    @classmethod
+    def create_sample_input(cls):
+        """
+        Needs to be implemented by the user in the inherited class.
+        Returns examples for data, model_config, samples_config.
+        This is useful for understanding the required 
+        data structures for the user model.
+        """
+        raise NotImplementedError
+
+    def build(self):
+        with self:
+            self._build()
+
+    def save(self, file_prefix, filepath, save_model=True, save_idata=True):
+        if save_idata:
+            file = Path(filepath + str(file_prefix) + ".nc")
+            self.idata.to_netcdf(file)
+        if save_model:
+            filepath = Path(str(filepath) + str(file_prefix) + ".pickle")
+            Model = cloudpickle.dumps(self)
+            file = open(filepath, "wb")
+            file.write(Model)
+        self.saved = True
+
+    @classmethod
+    def load(cls, file_prefix, filepath, load_model=True, laod_idata=True):
+        file = Path(str(filepath) + str(file_prefix) + ".pickle")
+        file = open(file, "rb")
+        self = cloudpickle.loads(file.read())
```

Comment:
A couple of comments here.

1. Wouldn't it be better to use a context manager to handle the connection? That makes sure the connection is closed, which is not happening here. 
2. You're loading a file and assigning it to `self` without knowing if the file you're reading contains an instance of `ModelBuilder`. This class may be helpful

```python
class Loader:
    def __init__(self):
        pass

    def save_to_pickle(self, filename):
        pickle_file = open(filename, "wb")
        pickle.dump(self, file=pickle_file)
        pickle_file.close()
        return None

    @classmethod
    def load_from_pickle(self, filename):
        with open(filename, "rb") as pickle_file:
            model = pickle.load(pickle_file)
        if isinstance(model, self):
            return model
        else:
            raise ValueError(
                f"The route '{filename}' does not contain an object of the class '{self.__name__}'"
            )
```

Then you can do

```python
class Model(Loader):
    def __init__(self):
        self.abc = 1
 
my_model = Model.load_from_pickle("file.pkl")
```

If you call `load_from_pickle` using an pkl that contains an object of a different class, it will fail. 
I'm not suggesting you to use the same names, but I do think this logic can be helpful

Reply:
Be aware of the limitations of `pickle`/`cloudpickle`. Quoting from [the `cloudpickle` README](https://github.com/cloudpipe/cloudpickle#cloudpickle):

> Cloudpickle can only be used to send objects between the **exact same version
of Python**.
>
> Using `cloudpickle` for **long-term object storage is not supported and
strongly discouraged.**
>
> **Security notice**: one should **only load pickle data from trusted sources** as
otherwise `pickle.load` can lead to arbitrary code execution resulting in a critical
security vulnerability.

Reply:
> @tomicapretto I'm curious about Bambi which shares similarities (intentionally). Do you think this would be useful for it?

What we don't have yet is a `.save()` method to save the model and/or the InferenceData object. But apart from that, most of the functionality is currently supported if I understand the aim of this. You can rebuild the Bambi model (which rebuilds the PyMC model) and load the NetCDF file with the InferenceData the model generated. This allows you to keep doing inferences later in time. The problem is, in Bambi, what if you load the wrong combination of model/NetCDF?  Sometimes it will fail, others it may silently do the wrong thing.

What would be better, at least on Bambi's side, is something that stores model configuration, the data, and the NetCDF.

Reply:
That's what the idea of the `id` hash is, even though we currently don't check it yet.